### PR TITLE
fix(tianmu): Avoid OOM when data too large and aggregated (#949)

### LINKED ISSUE #949: feature: Cache disk files for intermediate results of aggregate operations to avoid OOM
### Abstract:

TIANMU Engine - Aggregate operation intermediate results cache disk files to avoid OOM- requirements analysis

Related ISSUE: https://github.com/stoneatom/stonedb/issues/21

### Note of context:

The results of the current aggregate operation are cached in the in-memory HASH. Once the amount of data exceeds the RAM, the OOM occurs.

If the data volume exceeds RAM, convert it to disk file storage to avoid OOM

#### 1. The HASH result of the aggregation operation is cached in the disk file and properly participates in the operation

When the intermediate result exceeds RAM

#### 1. The compression ratio of the cache file size to the original data content, and the ratio of disk space to memory

Influence compression algorithm

#### 2. Disk cache file write speed and read speed

Standard disk hardware

It affects the speed of aggregation operations and the rules for writing and reading disk cache files

### Development cycle:

TODO:

### PR DESCRIPTION
**Pull request by adofsauron**
1. Use memory mapping
    2. Set the maximum number of concurrent threads for aggregation

## Summary about this PR

Issue Number: close #949  #1291

## Tests Check List

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog

- [ ] New Feature
- [x] Bug Fix
- [ ] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

### CODE REVIEW

File: storage/tianmu/mm/huge_heap_policy.cpp [hunk 1]
```diff
@@ -16,18 +16,30 @@
...
 
 namespace Tianmu {
 namespace mm {
 
 HugeHeap::HugeHeap(std::string hugedir, size_t size) : TCMHeap(0) {
+  {
+    char command[2048];
+    std::strcpy(command, "rm -rf ");
+    std::strcat(command, hugedir.c_str());
+    std::strcat(command, "/tianmuhuge.*");
```
Comment (wisehead) [inline storage/tianmu/mm/huge_heap_policy.cpp:39]: which directory does the huge files exist?

File: storage/tianmu/core/aggregation_algorithm.cpp [hunk 2]
```diff
@@ -247,8 +248,10 @@ void AggregationAlgorithm::MultiDimensionalGroupByScan(GroupByWrapper &gbw, int6
   unsigned int thd_cnt = 1;
   if (tianmu_sysvar_groupby_parallel_degree > 1) {
     if (static_cast<uint64_t>(mit.NumOfTuples()) > tianmu_sysvar_groupby_parallel_rows_minimum) {
-      unsigned int thd_limit = std::thread::hardware_concurrency() * 2;
+      unsigned int thd_limit = std::thread::hardware_concurrency();
+      thd_limit = thd_limit > 8 ? 8 : thd_limit;  // limit no more 8
       thd_cnt = tianmu_sysvar_groupby_parallel_degree > thd_limit ? thd_limit : tianmu_sysvar_groupby_parallel_degree;
+      thd_cnt = tianmu_sysvar_groupby_parallel_degree;
```
Comment (wisehead) [inline storage/tianmu/core/aggregation_algorithm.cpp]: thd_cnt is already set in above line.thd_cnt = tianmu_sysvar_groupby_parallel_degree > thd_limit ? thd_limit : tianmu_sysvar_groupby_parallel_degree;

why set it again?

File: storage/tianmu/core/aggregation_algorithm.cpp [hunk 3]
```diff
@@ -641,6 +660,32 @@ AggregaGroupingResult AggregationAlgorithm::AggregatePackrow(GroupByWrapper &gbw
...
+    if (mem_available) {
+      const auto mem_info = MemoryStatisticsOS::Instance()->GetMemInfo();
+      int64_t mem_available_chg = mem_info.mem_available - mem_available;
+      int64_t swap_used_chg = mem_info.swap_used - swap_used;
+
+      if (mem_used && (mem_available_chg < 0)) {
+        (*mem_used) -= mem_available_chg;
+      }
+
+      // #ifdef DEBUG_AGGREGA_PACK_COST
```
Comment (wisehead) [inline storage/tianmu/core/aggregation_algorithm.cpp]: remove this code if commented out.

File: storage/tianmu/handler/ha_tianmu.cpp [hunk 4]
```diff
@@ -2494,6 +2494,10 @@ static MYSQL_SYSVAR_UINT(insert_max_buffered, tianmu_sysvar_insert_max_buffered,
 static MYSQL_SYSVAR_BOOL(compensation_start, tianmu_sysvar_compensation_start, PLUGIN_VAR_BOOL, "-", nullptr, nullptr,
                          FALSE);
 static MYSQL_SYSVAR_STR(hugefiledir, tianmu_sysvar_hugefiledir, PLUGIN_VAR_READONLY, "-", nullptr, nullptr, "");
+static MYSQL_SYSVAR_UINT(os_least_mem, tianmu_os_least_mem, PLUGIN_VAR_READONLY, "-", nullptr, nullptr, 1, 0,
```
Comment (wisehead) [inline storage/tianmu/handler/ha_tianmu.cpp:2385]: what is the default value of os_least_mem?

File: storage/tianmu/handler/ha_tianmu.cpp [hunk 5]
```diff
@@ -2494,6 +2494,10 @@ static MYSQL_SYSVAR_UINT(insert_max_buffered, tianmu_sysvar_insert_max_buffered,
...
+                         UINT32_MAX, 0);
+static MYSQL_SYSVAR_UINT(hugefilesize, tianmu_sysvar_hugefilesize, PLUGIN_VAR_READONLY, "-", nullptr, nullptr, 1, 0,
```
Comment (wisehead) [inline storage/tianmu/handler/ha_tianmu.cpp:2387]: what's the default value?

File: storage/tianmu/mm/memory_statistics.h [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+    fd_self_statm = -1;
+  }
+
+ private:
+  int fd_self_statm = -1;
+  static MemoryStatisticsOS *instance;
+};
+
+#ifndef MEMORY_STATISTICS
+#define MEMORY_STATISTICS(model, deal)                                                                                \
```
Comment (wisehead) [inline storage/tianmu/mm/memory_statistics.h]: don't use MACRO here, not easy to debug.

**Review by wisehead — CHANGES_REQUESTED** [review]

File: storage/tianmu/handler/ha_tianmu.cpp (hunk 4 above)
Comment (adofsauron) [reply storage/tianmu/handler/ha_tianmu.cpp:2385]: 512MB at my.cnf.sample

File: storage/tianmu/mm/huge_heap_policy.cpp (hunk 1 above)
Comment (adofsauron) [reply storage/tianmu/mm/huge_heap_policy.cpp:39]: This is a parameter configuration,

```bash
# DIR
tianmu_hugefiledir='/tmp'

```

File: storage/tianmu/core/aggregation_algorithm.cpp (hunk 2 above)
Comment (adofsauron) [reply storage/tianmu/core/aggregation_algorithm.cpp]: Redundant code. Forgot to remove it

File: storage/tianmu/handler/ha_tianmu.cpp (hunk 5 above)
Comment (adofsauron) [reply storage/tianmu/handler/ha_tianmu.cpp:2387]: in the default my.cnf.sample
 

```

# DIR
tianmu_hugefiledir='/tmp'

# MB, If the value is set to 0, the system does not use it
tianmu_hugefilesize=30720

# minimum available memory of the system, MB
# If the remaining memory is smaller than this value, huge memory is used
tianmu_os_least_mem=512


```

**Review by wisehead — APPROVED** [review]

**Review by haitaoguan — APPROVED** [review]

**Pull request merged**